lib/OpenQA/Base.pm

```diff
@@ -13,20 +13,110 @@
 # You should have received a copy of the GNU General Public License along
 # with this program; if not, see <http://www.gnu.org/licenses/>.
 
-# This package contains shared functions between WebAPI and WebSockets
-package OpenQA::ServerStartup;
-
+package OpenQA::Base;
+use Mojo::Log;
+use Mojo::Home;
 use strict;
 use warnings;
+use Mojo::Base -base;
+use Sys::Hostname;
+use File::Spec::Functions 'catfile';
+use Mojo::File 'path';
 use Config::IniFiles;
 use db_profiler;
 use db_helpers;
 use OpenQA::Utils;
-use Mojo::File 'path';
 
-sub read_config {
-    my $app = shift;
+has config => sub { {} };
+
+has log => sub { setup_log(shift) };
+
+has home => sub { Mojo::Home->new($ENV{MOJO_HOME} || '/') };
+
+has mode => 'production';
+
+has log_name => 'scheduler';
+
+has level => 'debug';
+
+has instance => undef;
+
+has log_dir => undef;
+
+has schema => sub { OpenQA::Schema::connect_db() };
+
+sub setup_log {
+    my ($self) = @_;
+    my ($logfile, $logdir, $level, $log);
+
+    if ($self->isa('OpenQA::Base')) {
+        $logdir = $self->log_dir;
+        $level  = $self->level;
+    }
+    else {
+        $log = $self->log;
+        $level = $self->config->{logging}->{level} || 'debug';
+    }
+    $logfile = $ENV{OPENQA_LOGFILE} || $self->config->{logging}->{file};
+
+    if ($logfile && $logdir) {
+        $logfile = catfile($logdir, $logfile);
+        $log = Mojo::Log->new(
+            handle => path($logfile)->open('>>'),
+            level  => $self->level,
+            format => sub { return log_format($self->log_name, @_); });
+    }
+    elsif ($logfile) {
+        $log = Mojo::Log->new(
+            handle => path($logfile)->open('>>'),
+            level  => $level,
+            format => sub { return log_format($self->log_name, @_); });
+    }
+    elsif ($logdir) {
+        # So each worker from each host get it's own log (as the folder can be shared). Hopefully the machine hostname
+        # is already sanitized. Otherwise we need to check
+        $logfile
+          = catfile($logdir, hostname() . (defined $self->instance ? "-${\$self->instance}" : '') . ".log");
```

Comment:
just double check that the file will end up being unique and belonging to one instance (and that isn't checked waay to late)... perhaps starting two workers with the same instance number might give a hint. (Second worker should die due to the pool directory being locked).

Reply:
@foursixnine 

the call to hostname should ensure that there's no different workers (from different machines) with the same log file.

Reply:
Yeah, but I mean the case where there are two workers within the same machine, with the same instance number. 

simply on your local machine call `script/worker --verbose` twice and see the behaviour.

Reply:
I did and it gives this:
> /var/lib/openqa/pool/2 already locked

So it's not possible to launch two workers from the same machine